src/librustc_typeck/check/mod.rs

```diff
@@ -4955,10 +4955,46 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 Applicability::MachineApplicable,
             );
             err.note(
-                "for more on the distinction between the stack and the \
-                        heap, read https://doc.rust-lang.org/book/ch15-01-box.html, \
-                        https://doc.rust-lang.org/rust-by-example/std/box.html, and \
-                        https://doc.rust-lang.org/std/boxed/index.html",
+                "for more on the distinction between the stack and the heap, read \
+                 https://doc.rust-lang.org/book/ch15-01-box.html, \
+                 https://doc.rust-lang.org/rust-by-example/std/box.html, and \
+                 https://doc.rust-lang.org/std/boxed/index.html",
+            );
+        }
+    }
+
+    /// When encountering an `impl Future` where `BoxFuture` is expected, suggest `Box::pin`.
+    fn suggest_calling_boxed_future_when_appropriate(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        expr: &hir::Expr<'_>,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+    ) {
+        // Handle #68197.
+
+        if self.tcx.hir().is_const_context(expr.hir_id) {
+            // Do not suggest `Box::new` in const context.
+            return;
+        }
+        let pin_did = self.tcx.lang_items().pin_type();
+        match expected.kind {
+            ty::Adt(def, _) if Some(def.did) != pin_did => return,
+            // This guards the `unwrap` and `mk_box` below.
+            _ if pin_did.is_none() || self.tcx.lang_items().owned_box().is_none() => return,
+            _ => {}
+        }
+        let boxed_found = self.tcx.mk_box(found);
+        let new_found = self.tcx.mk_lang_item(boxed_found, lang_items::PinTypeLangItem).unwrap();
+        if let (true, Ok(snippet)) = (
+            self.can_coerce(new_found, expected),
+            self.sess().source_map().span_to_snippet(expr.span),
+        ) {
+            err.span_suggestion(
+                expr.span,
+                "you need to pin and box this expression",
+                format!("Box::pin({})", snippet),
```

Comment:
Instead of giving a suggestion that can mislead the user, can we only say `try using Box::pin()`?

context: https://github.com/rust-lang/rust/issues/68197#issuecomment-584964406

Reply:
It's not only the text of the suggestion that might mislead you, people can also skip all the way to the bottom `expected`/`found` note and think "oh, `Pin<Box<_>>` was expected, I'll just use `Pin::new(Box::new(_))`.

Reply:
Also, it wouldn't surprise me if people were writing `Box::new(async { .. })` proactively (not many, but enough people to want to deal with these cases).

Reply:
I think a significant number of people will proactively write `Box::new(..)`.

I understand that seeing `Pin<Box<_>>` is an existing problem, but it's not incorrect, and it's not the compiler actively telling you what to do.

With this change, we would be giving advice that is incorrect in a significant (albeit small) number of cases. I'd rather be slightly _less_ helpful, on average, if it means never being wrong. Bad advice from the compiler can lead people to disregard diagnostic hints altogether.

So my suggestion is to _point_ the user to `Box::pin` with `consider using Box::pin`, but not actually show them how to use it (unless we know the code sample is correct when we do). The user is free to guess how to use it, and if that fails, look it up the docs. But this avoids the exasperation / despair of "I did _exactly_ what the compiler said and it still doesn't work."

Reply:
What do you think of the new output?

https://github.com/rust-lang/rust/pull/69082/files#diff-634cabb162689069280991095dd769cd

Reply:
Looks good to me :)